hack/fake-apiserver/main.go

```diff
@@ -0,0 +1,312 @@
+package main
+
+import (
+	"context"
+	"crypto/rsa"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	infrav1 "sigs.k8s.io/cluster-api/test/infrastructure/inmemory/api/v1alpha1"
+
+	"encoding/base64"
+	"encoding/json"
+
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	cmanager "sigs.k8s.io/cluster-api/test/infrastructure/inmemory/pkg/runtime/manager"
+	"sigs.k8s.io/cluster-api/test/infrastructure/inmemory/pkg/server"
+	"sigs.k8s.io/cluster-api/util/certs"
+)
+
+var (
+	cloudScheme  = runtime.NewScheme()
+	setupLog     = ctrl.Log.WithName("setup")
+	scheme       = runtime.NewScheme()
+	cloudMgr     = cmanager.New(cloudScheme)
+	apiServerMux = &server.WorkloadClustersMux{}
+	key          *rsa.PrivateKey
+	ctx          = context.Background()
+)
+
+func init() {
+	// scheme used for operating on the management cluster.
+	_ = clientgoscheme.AddToScheme(scheme)
+	_ = clusterv1.AddToScheme(scheme)
+	_ = infrav1.AddToScheme(scheme)
+
+	// scheme used for operating on the cloud resource.
+	_ = corev1.AddToScheme(cloudScheme)
+	_ = appsv1.AddToScheme(cloudScheme)
+	_ = rbacv1.AddToScheme(cloudScheme)
+}
+
+type ResourceData struct {
+	ResourceName string
+	Host         string
+	Port         int
+}
+
+func register(w http.ResponseWriter, r *http.Request) {
```

Comment:
Could you please add a docstring here describing what this function does? Include the URL parameters